server/src/main/java/org/opensearch/wlm/cancellation/TaskCanceller.java

```diff
@@ -0,0 +1,244 @@
+ * This class utilizes a strategy pattern through {@link MaximumResourceTaskSelectionStrategy} to identify tasks that exceed
+ * predefined resource usage limits and are therefore eligible for cancellation.
+ *
+ * <p>The cancellation process is initiated by evaluating the resource usage of each QueryGroup against its
+ * resource limits. Tasks that contribute to exceeding these limits are selected for cancellation based on the
+ * implemented task selection strategy.</p>
+ *
+ * <p>Instances of this class are configured with a map linking QueryGroup IDs to their corresponding resource usage
+ * views, a set of active QueryGroups, and a task selection strategy. These components collectively facilitate the
+ * identification and cancellation of tasks that threaten to breach QueryGroup resource limits.</p>
+ *
+ * @see MaximumResourceTaskSelectionStrategy
+ * @see QueryGroup
+ * @see ResourceType
+ */
+public class TaskCanceller {
+    public static final double MIN_VALUE = 1e-9;
+
+    private final WorkloadManagementSettings workloadManagementSettings;
+    private final TaskSelectionStrategy taskSelectionStrategy;
+    // a map of QueryGroupId to its corresponding QueryGroupLevelResourceUsageView object
+    private final Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews;
+    private final Collection<QueryGroup> activeQueryGroups;
+    private final Collection<QueryGroup> deletedQueryGroups;
+    private BooleanSupplier isNodeInDuress;
+
+    public TaskCanceller(
+        WorkloadManagementSettings workloadManagementSettings,
+        MaximumResourceTaskSelectionStrategy taskSelectionStrategy,
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews,
+        Collection<QueryGroup> activeQueryGroups,
+        Collection<QueryGroup> deletedQueryGroups,
+        BooleanSupplier isNodeInDuress
+    ) {
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.taskSelectionStrategy = taskSelectionStrategy;
+        this.queryGroupLevelResourceUsageViews = queryGroupLevelResourceUsageViews;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        this.isNodeInDuress = isNodeInDuress;
+    }
+
+    /**
+     * Cancel tasks based on the implemented strategy.
+     */
+    public final void cancelTasks() {
+        // cancel tasks from QueryGroups that are in Enforced mode that are breaching their resource limits
+        cancelTasks(ResiliencyMode.ENFORCED);
+        // if the node is in duress, cancel tasks accordingly.
+        handleNodeDuress();
+    }
+
+    private void handleNodeDuress() {
+        if (!isNodeInDuress.getAsBoolean()) {
+            return;
+        }
+        // List of tasks to be executed in order if the node is in duress
+        List<Consumer<Void>> duressActions = List.of(v -> cancelTasksFromDeletedQueryGroups(), v -> cancelTasks(ResiliencyMode.SOFT));
```

Comment:
What is the purpose of `cancelTasksFromDeletedQueryGroups`?

Reply:
We should prioritise deleting the tasks from deleted query groups first if the node is in duress. Deletion happens for the following order
* Enforced always
* When node is in duress 
    1. Deletedfrom deleted query groups
    2. Delete from SOFT mode query groups